src/infra/dex/balancer/dto.rs

```diff
@@ -1,58 +1,208 @@
+            assetOutIndex
+            amount
+            userData
+        }
+        swapAmountRaw
+        returnAmountRaw
+        tokenIn
+        tokenOut
+    }
+}
+    "#;
+}
+
+#[derive(Serialize)]
 #[serde(rename_all = "camelCase")]
-pub enum OrderKind {
-    Sell,
-    Buy,
+pub struct Query<'a> {
+    query: &'a str,
+    variables: Variables,
+}
+
+impl Query<'_> {
+    pub fn from_domain(
+        order: &dex::Order,
+        slippage: &dex::Slippage,
+        chain_id: eth::ChainId,
+        contract_address: eth::ContractAddress,
+        query_batch_swap: bool,
+    ) -> Result<Self, Error> {
+        let variables = Variables {
+            call_data_input: CalDataInput {
+                deadline: None,
+                receiver: contract_address.0,
+                sender: contract_address.0,
+                slippage_percentage: slippage.as_factor().clone(),
+            },
+            chain: Chain::from_domain(chain_id)?,
+            query_batch_swap,
+            swap_amount: EtherAmount::from_wei(&order.amount.get()),
+            swap_type: SwapType::from_domain(order.side),
+            token_in: order.sell.0,
+            token_out: order.buy.0,
+            use_vault_version: VaultVersion::V2 as u8,
+        };
+        Ok(Self {
+            query: get_swap_paths_query::QUERY,
+            variables,
+        })
+    }
 }
 
-/// An SOR query.
+/// Ether amount refers to the SOR API V3's `AmountHumanReadable` type.
 #[serde_as]
-#[derive(Debug, Serialize)]
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct EtherAmount(BigDecimal);
+
+impl EtherAmount {
+    pub fn from_wei(wei: &U256) -> EtherAmount {
+        Self(u256_to_big_decimal(wei) / BigDecimal::from(10_u64.pow(18)))
+    }
+
+    pub fn value(&self) -> BigDecimal {
+        self.0.clone()
+    }
+
+    pub fn to_wei(&self) -> Option<U256> {
+        big_decimal_to_u256(&(&self.0 * BigDecimal::from(10_u64.pow(18))))
+    }
+}
+
+#[serde_as]
+#[derive(Serialize)]
 #[serde(rename_all = "camelCase")]
-pub struct Query {
-    /// The sell token to quote.
-    pub sell_token: H160,
-    /// The buy token to quote.
-    pub buy_token: H160,
-    /// The order kind to use.
-    pub order_kind: OrderKind,
-    /// The amount to quote
-    ///
-    /// For sell orders this is the exact amount of sell token to trade, for buy
-    /// orders, this is the amount of buy tokens to buy.
-    #[serde_as(as = "serialize::U256")]
-    pub amount: U256,
-    /// The current gas price estimate used for determining how the trading
-    /// route should be split.
-    #[serde_as(as = "serialize::U256")]
-    pub gas_price: U256,
+struct Variables {
+    call_data_input: CalDataInput,
+    /// The Chain to query.
+    chain: Chain,
+    /// Whether to run `queryBatchSwap` to update the return amount with most
+    /// up-to-date on-chain values.
+    query_batch_swap: bool,
+    /// The amount to swap in human form.
+    swap_amount: EtherAmount,
```

Comment:
A drawback of the new API is that an Ether value should be used in the request instead of WEI.

Reply:
Is this actually an ether amount or just the token amount divided by decimals?

Reply:
Right, that is just an amount divided by decimals. I just realized this wouldn't work with tokens that do not have 18 decimals. The decimals amount should be either fetched or provided via the order struct, right?